File: Data Structures & Algorithms/combination-target-sum/submission-2.cpp
class Solution {
    void backtrack(vector<int>& nums,int target,vector<int>& ans,int index) {
        if(target < 0 || index >= nums.size()) return;
        if(target == 0) {
            res.push_back(ans);
            return;
        }

        if(target > 0) backtrack(nums,target,ans,index+1);
        if(target - nums[index] >= 0) {
            ans.push_back(nums[index]);
            backtrack(nums,target-nums[index],ans,index);
            ans.pop_back();
        }
        return;
    }
public:
    vector<vector<int>> res;
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int> ans;
        backtrack(nums,target,ans,0);
        return res;
    }
};

File: Data Structures & Algorithms/search-in-rotated-sorted-array-ii/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        l, r = 0, len(nums)-1
        while l <= r:
            mid = l + (r-l)//2
            if nums[mid] == target:
                return True
            elif nums[l] < nums[mid]:
                if target < nums[mid] and nums[l] <= target:
                    r = mid - 1
                else:
                    l = mid + 1
            elif nums[l] > nums[mid]:
                if nums[mid] < target and target <= nums[r]:
                    l = mid + 1
                else:
                    r = mid - 1
            else:
                l += 1
        
        return False